src/Npgsql/PostgresDatabaseInfo.cs

```diff
@@ -201,161 +203,139 @@ internal async Task<List<PostgresType>> LoadBackendTypes(NpgsqlConnector conn, N
             if (timeout.IsSet)
                 commandTimeout = (int)timeout.CheckAndGetTimeLeft().TotalSeconds;
 
-            using var batch = new NpgsqlBatch(conn)
-            {
-                BatchCommands =
-                {
-                    new("SELECT version()"),
-                    new(GenerateLoadTypesQuery(SupportsRangeTypes, conn.Settings.LoadTableComposites)),
-                    new(GenerateLoadCompositeTypesQuery(conn.Settings.LoadTableComposites)),
-                }
-            };
+            var batchQuery = new StringBuilder("SELECT version();")
+                .Append(GenerateLoadTypesQuery(SupportsRangeTypes, conn.Settings.LoadTableComposites))
+                .Append(GenerateLoadCompositeTypesQuery(conn.Settings.LoadTableComposites));
 
             if (SupportsEnumTypes)
-                batch.BatchCommands.Add(new(GenerateLoadEnumFieldsQuery(HasEnumSortOrder)));
-
-            batch.Timeout = commandTimeout;
-            batch.AllResultTypesAreUnknown = true;
+                batchQuery.Append(GenerateLoadEnumFieldsQuery(HasEnumSortOrder));
 
+            conn.WriteBuffer.Timeout = TimeSpan.FromSeconds(commandTimeout);
             timeout.CheckAndApply(conn);
-            var reader = async ? await batch.ExecuteReaderAsync() : batch.ExecuteReader();
-            try
-            {
-                var byOID = new Dictionary<uint, PostgresType>();
 
-                // First the PostgreSQL version
-                if (async)
-                {
-                    await reader.ReadAsync();
-                    LongVersion = reader.GetString(0);
-                    await reader.NextResultAsync();
-                }
-                else
+            var q = batchQuery.ToString();
+            Debug.WriteLine($"Type loading query:{Environment.NewLine}{q}");
+            await conn.WriteQuery(q, async);
+            await conn.Flush(async);
+            var byOID = new Dictionary<uint, PostgresType>();
+            var buf = conn.ReadBuffer;
+
+            // First read the PostgreSQL version
+            var rowDescription = Expect<RowDescriptionMessage>(await conn.ReadMessage(async), conn);
```

Comment:
Too bad things are slightly different with the extended protocol (e.g. no ParseCompleteMessage...). It would be nice to have a code path that produces a DbDataReader at some point (but of course this looks good for now).

Reply:
Yeah, the current implementation is really pretty raw but I actually expected it to be worse.